src/predictor/cpu_predictor.cc

```diff
@@ -171,18 +171,25 @@ class CPUPredictor : public Predictor {
     if (ntree_limit == 0 || ntree_limit > model.trees.size()) {
       ntree_limit = static_cast<unsigned>(model.trees.size());
     }
-
+    std::cout << "run PredLoopInternal in PredictBatch\n";
     this->PredLoopInternal(dmat, &out_preds->HostVector(), model,
                            tree_begin, ntree_limit);
 
     auto cache_entry = this->FindCache(dmat);
     if (cache_entry == cache_->cend()) {
-      return;
+      std::cout << "cannot find cache\n";
+      if (!generic_param_->adding_all_to_cache || !(*cache_).empty()) {
+        return;
+      } else {
+        std::cout << "adding dmatrix to cache\n";
+        (*cache_)[dmat].data = static_cast<std::shared_ptr<DMatrix>>(dmat);
```

Comment:
I am a bit confused by this line of code. It was quite a while when I last dived into this part of code
(Full disclaimer, I recently switched companies and I am no longer work in Criteo on xgboost :( )

My C++ is a little bit rusty in regard to static_cast in this context so please feel free to correct me
I am reading this https://en.cppreference.com/w/cpp/language/static_cast and I am trying to figure out which case is here.
A constructor of shared_ptr is marked as explicit so it can't be an implicit object creation but I do not understand then what it can be? shared_ptr<DMatrix> is not a child of DMatrix.

Are you creating a new shared_ptr from the pointer dmat? If so I strongly suggest against. It usually leads to errors.

Cause I thought about similar fix but I rejected it cause I did not figure out how to retrieve already existing shared_pt<DMatrix> from just DMatrix* (cause I want to have the copy of that reference counting pointer)

What I had in mind is changing the interface of Predict to pass shared_ptr in the first place instead of just a raw pointer. That would permit us to recreate a cache

Reply:
I consulted with my C++ friends and they pointed to https://stackoverflow.com/questions/32713083/explicit-constructor-and-static-cast
So static_cast here actually will call a constructor and it will create a second separate shared_ptr which won't know about any other shared_ptr's which manage this pointer.

Sooner or later this will result in double free cause each of these smart pointers will try to free the resource.

 In the end I suggest to wait @trivialfis refactoring. It should enable you (or us) to solve this problem

Reply:
yeah that's why the current unit test broken